### PR TITLE
portFFT integration

### PR DESCRIPTION
**Pull request by FMarno**
# Description
Introduces the portFFT backend, a pure SYCL implemenation.

# Checklist

## All Submissions

- [ ] Do all unit tests pass locally? Attach a log.
- [ ] Have you formatted the code using clang-format?

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -63,6 +63,28 @@ option(ENABLE_ROCSOLVER_BACKEND "Enable the rocSOLVER backend for the LAPACK int
...
+endif()
+if (ENABLE_PORTFFT_BACKEND_AMD_GPU AND ENABLE_ROCFFT_BACKEND)
+  message(SEND_ERROR "ENABLE_PORTFFT_BACKEND_AMD_GPU and ENABLE_ROCFFT_BACKEND cannot be enabled at the same time. \
+  Only one DFT backend can be enable for each device type.")
+endif()
+if (ENABLE_PORTFFT_BACKEND_NVIDIA_GPU AND ENABLE_CUFFT_BACKEND)
+  message(SEND_ERROR "ENABLE_PORTFFT_BACKEND_NVIDIA_GPU and ENABLE_CUFFT_BACKEND cannot be enabled at the same time. \
+  Only one DFT backend can be enable for each device type.")
+endif()
+set(ENABLE_PORTFFT_BACKEND (ENABLE_PORTFFT_BACKEND_AMD_GPU || ENABLE_PORTFFT_BACKEND_INTEL_CPU ||ENABLE_PORTFFT_BACKEND_INTEL_GPU ||ENABLE_PORTFFT_BACKEND_NVIDIA_GPU))
```
Comment (Rbiessy) [inline CMakeLists.txt]: We don't need all of this logic for portFFT. I think it would be simpler if we only have `ENABLE_PORTFFT_BACKEND` (and same comment for the compiler defines).
I suppose this could let users combine portfft with other backends for different devices but the original idea was that portFFT should be enabled for all or nothing since the whole point is that it is portable. For portBLAS we check that the portBLAS backend cannot be used with other backends [here](https://github.com/oneapi-src/oneMKL/blob/develop/CMakeLists.txt#L118).

File: examples/README.md [hunk 2]
```diff
@@ -455,4 +480,29 @@ Device name is: AMD Radeon PRO W6800
...
+# SYCL device
+#
+########################################################################
+
+Running DFT complex forward example on GPU device
+Device name is: Intel(R) UHD Graphics 750
+Running with single precision real data type:
+DFT example run_time dispatch
+Caught std::exception:
+	oneMKL: dft/backends/portfft/commit: function is not implemented portFFT only supports complex to complex transforms
```
Comment (Rbiessy) [inline examples/README.md:483]: Is it really useful to add an example where portFFT does not support the configuration? I don't really get the point of adding example output to begin with.

File: src/dft/backends/portfft/commit.cpp [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+private:
+    using scalar_type = typename dft::detail::commit_impl<prec, dom>::scalar_type;
+    using fwd_type = typename dft::detail::commit_impl<prec, dom>::fwd_type;
+    using bwd_type = typename dft::detail::commit_impl<prec, dom>::bwd_type;
+    using descriptor_type = typename dft::detail::descriptor<prec, dom>;
+
+    static constexpr pfft::domain domain =
+        std::conditional_t<dom == dft::domain::REAL,
+                           std::integral_constant<pfft::domain, pfft::domain::REAL>,
+                           std::integral_constant<pfft::domain, pfft::domain::COMPLEX>>::value;
```
Comment (Rbiessy) [inline src/dft/backends/portfft/commit.cpp]: Isn't this equivalent to the following?

```suggestion
    static constexpr pfft::domain domain = dom == dft::domain::REAL ? pfft::domain::REAL : pfft::domain::COMPLEX;
```

File: src/dft/backends/portfft/commit.cpp [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        fwd_desc.placement = config_values.placement == config_value::INPLACE
+                                 ? pfft::placement::IN_PLACE
+                                 : pfft::placement::OUT_OF_PLACE;
+        fwd_desc.forward_offset = static_cast<std::size_t>(config_values.input_strides[0]);
+        fwd_desc.backward_offset = static_cast<std::size_t>(config_values.output_strides[0]);
+        fwd_desc.forward_strides = { config_values.input_strides.cbegin() + 1,
+                                     config_values.input_strides.cend() };
+        fwd_desc.backward_strides = { config_values.output_strides.cbegin() + 1,
+                                      config_values.output_strides.cend() };
+        // guard against overwriting the default value when not needed, this seems to lead to issues.
```
Comment (Rbiessy) [inline src/dft/backends/portfft/commit.cpp]: Can we investigate more what the issue is? portFFT shouldn't care about the distances set when the number of transforms is 1.

File: tests/unit_tests/CMakeLists.txt [hunk 5]
```diff
@@ -106,65 +106,57 @@ foreach(domain ${TARGET_DOMAINS})
...
     list(APPEND ONEMKL_LIBRARIES_${domain} onemkl_${domain}_rocrand)
   endif()
 
   if(domain STREQUAL "dft" AND ENABLE_CUFFT_BACKEND)
-    add_dependencies(test_main_${domain}_ct onemkl_${domain}_cufft)
     list(APPEND ONEMKL_LIBRARIES_${domain} onemkl_${domain}_cufft)
   endif()
 
   if(domain STREQUAL "dft" AND ENABLE_ROCFFT_BACKEND)
-    add_dependencies(test_main_${domain}_ct onemkl_dft_rocfft)
```
Comment (Rbiessy) [inline tests/unit_tests/CMakeLists.txt]: I'd rather that we don't remove these dependencies as part of this PR and follow the same format for portfft. If we really want to improve this it should be discussed in a different PR IMO.

File: tests/unit_tests/dft/include/compute_tester.hpp [hunk 6]
```diff
@@ -51,12 +51,12 @@ struct DFT_Test {
...
-    const std::vector<std::int64_t> strides_bwd;
-    const std::int64_t batches;
-    const std::int64_t forward_elements;
-    const std::size_t size_total;
+    std::vector<std::int64_t> sizes;
+    std::vector<std::int64_t> strides_fwd;
+    std::vector<std::int64_t> strides_bwd;
+    std::int64_t batches;
+    std::int64_t forward_elements;
+    std::size_t size_total;
```
Comment (Rbiessy) [inline tests/unit_tests/dft/include/compute_tester.hpp]: Why not keeping them `const`?

Comment (FMarno) [reply tests/unit_tests/dft/include/compute_tester.hpp]: `strides_fwd` and `strides_bwd` were shadowed in `compute_inplace.hpp` so they could be modified, and this allowed me to just modify them in-place, but I decided it's not worth fixing right now so have reverted it here https://github.com/FMarno/oneMKL/pull/1/commits/d2fb2bc96d8b5fe66d4e34b744ede1ecda953710

File: tests/unit_tests/CMakeLists.txt (hunk 5 above)
Comment (FMarno) [reply tests/unit_tests/CMakeLists.txt]: ok, reverted here https://github.com/FMarno/oneMKL/pull/1/commits/d2fb2bc96d8b5fe66d4e34b744ede1ecda953710

File: src/dft/backends/portfft/commit.cpp (hunk 3 above)
Comment (FMarno) [reply src/dft/backends/portfft/commit.cpp]: Yes! I guess I overcomplicated because templates metaprogramming can't be that easy :sweat_smile: 
fixed in https://github.com/FMarno/oneMKL/pull/1/commits/fc720ead06549335a71310bb2dc44e976703488b

File: src/dft/backends/portfft/commit.cpp (hunk 4 above)
Comment (FMarno) [reply src/dft/backends/portfft/commit.cpp]: I checked and this has been fixed since I first added the guard https://github.com/FMarno/oneMKL/pull/1/commits/fc720ead06549335a71310bb2dc44e976703488b

File: examples/README.md (hunk 2 above)
Comment (FMarno) [reply examples/README.md:483]: I was kind of just following what the other backends had done. I supposed the example output might be helpful to check that you are seeing the correct thing.

I could just not have the example. I'm sure uses can figure it out from existing examples.

Comment (FMarno) [reply examples/README.md:483]: actually, this is the runtime dispatch example so I don't want to modify that example. 
All the other backends have their output here, so maybe it's silly but it's honest

Comment (FMarno) [reply examples/README.md:483]: comptime example removed here: https://github.com/FMarno/oneMKL/pull/1/commits/fb0ad2f9f2d9c9b8aa630c3eaa2b58dee04f5d60

File: CMakeLists.txt (hunk 1 above)
Comment (FMarno) [reply CMakeLists.txt]: Addressed in https://github.com/FMarno/oneMKL/pull/1/commits/57e39ab53bb2a6d15dd4a9592f11517907953d53

File: CMakeLists.txt [hunk 7]
```diff
@@ -102,22 +103,31 @@ endif()
...
+		OR ENABLE_NETLIB_BACKEND))
   message(FATAL_ERROR "portBLAS backend cannot be enabled with other backends")
 endif()
 
+if (ENABLE_PORTFFT_BACKEND
+	AND (ENABLE_MKLCPU_BACKEND
+		OR ENABLE_MKLGPU_BACKEND
+		OR ENABLE_ROCFFT_BACKEND
+		OR ENABLE_CUFFT_BACKEND))
+  message(SEND_ERROR "ENABLE_PORTFFT_BACKEND cannot be enabled at the same time as other DFT backends.")
```
Comment (Rbiessy) [inline CMakeLists.txt]: I think we want to stop processing if that happens so it should be a `FATAL_ERROR`. Can we also make the message consistent with portBLAS above (whichever we end up using)?

File: examples/README.md (hunk 2 above)
Comment (Rbiessy) [reply examples/README.md:483]: Yeah we don't need an example for every backend. That's fine, I'll make a note to update this example once portFFT supports real in oneMKL.

File: tests/unit_tests/dft/include/compute_tester.hpp (hunk 6 above)
Comment (Rbiessy) [reply tests/unit_tests/dft/include/compute_tester.hpp]: Oh ok, I think I remember seeing warnings because of unused `strides_fwd` and `strides_bwd` now. Do you see these warnings or are they fixed already?

File: tests/unit_tests/CMakeLists.txt (hunk 5 above)
Comment (Rbiessy) [reply tests/unit_tests/CMakeLists.txt]: How about we also add the dependency for portFFT backend to keep it consistent for now?

File: tests/unit_tests/dft/include/compute_tester.hpp (hunk 6 above)
Comment (FMarno) [reply tests/unit_tests/dft/include/compute_tester.hpp]: I have left the warnings as they are

File: CMakeLists.txt (hunk 7 above)
Comment (FMarno) [reply CMakeLists.txt]: That makes sense https://github.com/FMarno/oneMKL/pull/1/commits/516210b81b221a786cedbc239d76b477eb11756b

File: tests/unit_tests/CMakeLists.txt (hunk 5 above)
Comment (FMarno) [reply tests/unit_tests/CMakeLists.txt]: done https://github.com/FMarno/oneMKL/pull/1/commits/a5480c9781e8242675be27086ded0881bb6fda0e

File: tests/unit_tests/dft/include/compute_tester.hpp (hunk 6 above)
Comment (Rbiessy) [reply tests/unit_tests/dft/include/compute_tester.hpp]: Oh, I think it would be nice to fix all the warnings we see when enabling this backend. I didn't realise this was your suggestion to fix warnings. It would make sense to revert to your first version I think!

Comment (FMarno) [reply tests/unit_tests/dft/include/compute_tester.hpp]: Ok, I've also update a small piece of code to hopefully make it clearer. https://github.com/FMarno/oneMKL/pull/1/commits/fda4b7842e0be270b8621640e11e0183a3879a47

Also you're not supposed to have const members since it can prevent them being moved, so there's another reason.

**Comment by Rbiessy** [comment]
LGTM

**Comment by FMarno** [comment]
merged upstream
